File: leetcode/design/588_Design_In-Memory_File_System/codes/design.cpp
class FileSystem {
    struct Node {
        bool isFile = false;
        string content;
        map<string, Node*> children;
    };
    
    Node* root;

public:
    FileSystem() {
        root = new Node();
    }

    vector<string> ls(string path) {
        Node* node = traverse(path);
        if (node -> isFile) return {path.substr(path.find_last_of('/') + 1)};
        vector<string> res;
        for (auto& p : node -> children) res.push_back(p.first);
        return res;
    }

    void mkdir(string path) {
        traverse(path, true);
    }

    void addContentToFile(string filePath, string content) {
        Node* node = traverse(filePath, true);
        node -> isFile = true;
        node -> content += content;
    }

    string readContentFromFile(string filePath) {
        Node* node = traverse(filePath);
        return node -> content;
    }

private:
    Node* traverse(const string& path, bool create = false) {
        Node* node = root;
        stringstream ss(path);
        string part;
        getline(ss, part, '/'); 
        while (getline(ss, part, '/')) {
            if (!node -> children.count(part))
                node -> children[part] = new Node();
            node = node -> children[part];
        }
        return node;
    }
};

/**
 * Your FileSystem object will be instantiated and called as such:
 * FileSystem* obj = new FileSystem();
 * vector<string> param_1 = obj -> ls(path);
 * obj -> mkdir(path);
 * obj -> addContentToFile(filePath,content);
 * string param_4 = obj -> readContentFromFile(filePath);
 */

File: leetcode/design/588_Design_In-Memory_File_System/codes/design.py
class Node:
    def __init__(self):
        self.children = {}  
        self.is_file = False
        self.content = ""

class FileSystem:

    def __init__(self):
        self.root = Node()

    def ls(self, path: str):
        node = self._traverse(path)
        if node.is_file:
            return [path.split("/")[-1]]
        return sorted(node.children.keys())

    def mkdir(self, path: str):
        self._traverse(path, create=True)

    def addContentToFile(self, filePath: str, content: str):
        node = self._traverse(filePath, create=True)
        node.is_file = True
        node.content += content

    def readContentFromFile(self, filePath: str) -> str:
        node = self._traverse(filePath)
        return node.content

    def _traverse(self, path: str, create=False):
        node = self.root
        if path == "/":
            return node
        parts = path.split("/")
        for part in parts[1:]:
            if part not in node.children:
                if create:
                    node.children[part] = Node()
                else:
                    return None
            node = node.children[part]
        return node

# Your FileSystem object will be instantiated and called as such:
# obj = FileSystem()
# param_1 = obj.ls(path)
# obj.mkdir(path)
# obj.addContentToFile(filePath,content)
# param_4 = obj.readContentFromFile(filePath)